src/os/memstore/MemStore.h

```diff
@@ -195,7 +196,7 @@ class MemStore : public ObjectStore {
 
   Finisher finisher;
 
-  uint64_t used_bytes;
+  std::atomic<uint64_t> used_bytes;
```

Comment:
i cannot find any root cause analysis in the tracker ticket or in the commit message. according to the log line
```
2021-06-03T11:55:42.193+0000 7f815da76700 10 memstore(/var/lib/ceph/osd/ceph-4) statfs: used_bytes: 18446744073708990649/200000000
```

i think it was caused by integer underflow.

Reply:
I have count all the write size and all the remove size, they are all matched.

Seems the write and remove could be running in parallel, just in case:

used_bytes = 1001

    Thread_A                                  Thread_B
used_bytes += 1                     used_bytes -= 1000

If it expects the Thread_A to run first, then the result should be 1001 + 1 - 1000 = 2.
But in theory the Thread_B could be finish faster, then the result will be 1, so this will cause the underflow.

???

Reply:
how come the order matters? the results of
```py
used_bytes = 1001
used_bytes -= 1000
used_bytes += 1
```

and 
```py
used_bytes = 1001
used_bytes += 1
used_bytes -= 1000
```
are identical.

Reply:
Sorry for confusing, I meant for none atomic `used_bytes += XXX` will be translate to 3 assembly language commands like:

         Thread_A                                                                                                  Thread_B
A1, load(used_bytes from memory) to register eax                                B1, load(used_bytes from memory) to register ebx                          
A2, eax = eax + 1                                                                                    B2, ebx = ebx - 1000
A3, store(eax) to used_bytes in memory                                                      Nope
                                                                                                                B3, store(ebx) to used_bytes in memory

Reply:
i see. that makes sense. without mutating it in an atomic op or a compare-and-swap loop, we can not ensure the atomicity of the fetch-and-add operation, if the compiler choose to implement it. so the change per-se is correct, i think.

but i don't understand how it addresses the integer underflow issue. in the example above. `used_bytes` would be "1", which is wrong, but it does not underflow.

Reply:
For example:

At beginning the `used_bytes` is 0:

```
Thread_A                                Thread_B

A1,  used_bytes += 3
A2,  used_bytes += 4
A3,  used_bytes += 9             B1,  used_bytes -= 4 (4 is from step A2 in Thread_A)

A4,  used_bytes += 10

For the above case, after the step `A3` and `B1`, the `used_bytes` could be `(7+9) == 16` or `(7-4) == 3`. 

No matter which one it's buggy. For the later result, when all the objects are being delete, the `used_bytes -= ( 3 + 9 + 10)`, which is `3 - 22 == -19`.

Reply:
From the code and logs I didn't find any other suspicious case could cause the `used_bytes` underflow.